spec/ics-4-channel-and-packet-semantics/README.md

```diff
@@ -0,0 +1,757 @@
+to acknowledge the channel close request and continue the closing process.
+
+```typescript
+interface ChanCloseTry {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  timeoutHeight: uint64
+  nextTimeoutHeight: uint64
+  proofInit: CommitmentProof
+}
+```
+
+```typescript
+function chanCloseTry(
+  connectionIdentifier: Identifier, channelIdentifier: Identifier,
+  timeoutHeight: uint64, nextTimeoutHeight: uint64, proofInit: CommitmentProof) {
+  assert(getConsensusState().getHeight() < timeoutHeight)
+  channel = get(channelKey(connectionIdentifier, channelIdentifier))
+  assert(channel.state === OPEN)
+  connection = get("connections/{connectionIdentifier}")
+  assert(connection.state === OPEN)
+  consensusState = get("clients/{connection.clientIdentifier}/consensusState")
+  expected = Channel{INIT, channel.counterpartyModuleIdentifier, channel.moduleIdentifier,
+                     channel.channelIdentifier, timeoutHeight}
+  assert(verifyMembership(
+    consensusState,
+    proofInit,
+    "connections/{connection.counterpartyIdentifier}/channels/{channel.counterpartyChannelIdentifier}",
+    expected
+  ))
+  channel.state = CLOSED
+  channel.nextTimeoutHeight = nextTimeoutHeight
+  set(channelKey(connectionIdentifier, channelIdentifier), channel)
+}
+```
+
+The `chanCloseAck` function is called by the handshake-originating module
+to acknowledge the closing acknowledgement and finalize channel closure.
+
+```typescript
+interface ChanCloseAck {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  timeoutHeight: uint64
+  proofTry: CommitmentProof
+}
+```
+
+```typescript
+function chanCloseAck(
+  connectionIdentifier: Identifier, channelIdentifier: Identifier,
+  timeoutHeight: uint64, proofTry: CommitmentProof) {
+  assert(getConsensusState().getHeight() < timeoutHeight)
+  channel = get(channelKey(connectionIdentifier, channelIdentifier))
+  assert(channel.state === OPEN)
+  connection = get("connections/{connectionIdentifier}")
+  assert(connection.state === OPEN)
+  consensusState = get("clients/{connection.clientIdentifier}/consensusState")
+  expected = Channel{CLOSED, channel.counterpartyModuleIdentifier, channel.moduleIdentifier,
+                     channelIdentifier, timeoutHeight}
+  assert(verifyMembership(
+    consensusState.getRoot(),
+    proofInit,
+    channelKey(connection.counterpartyIdentifier, channel.counterpartyChannelIdentifier),
+    expected
+  ))
+  channel.state = CLOSED
+  channel.nextTimeoutHeight = 0
+  set(channelKey(connectionIdentifier, channelIdentifier), channel)
+}
+```
+
+The `chanCloseTimeout` function is called by either the handshake-originating
+or handshake-accepting module to prove a timeout and reset state.
+
+```typescript
+interface ChanCloseTimeout {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  timeoutHeight: uint64
+  proofTimeout: CommitmentProof
+}
```

Comment:
Maybe stupid question (I've might have missed a few discussions): What is with all these "interfaces"? They don't seem to contain any further information besides all parameters of the function (which are repeated below).

Reply:
No, it's a good question. They were from a previous incarnation of the model where the IBC module accepted datagrams directly, and I left them since an IBC relayer module might do the same (in ICS 26).

Now I think they may just be confusing though (as evidenced).

Reply:
Going to handle in a followup - https://github.com/cosmos/ics/issues/118